Firestore/core/src/firebase/firestore/model/base_path.h

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_BASE_PATH_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_BASE_PATH_H_
+
+#include <algorithm>
+#include <cctype>
+#include <initializer_list>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "Firestore/core/src/firebase/firestore/util/firebase_assert.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+namespace impl {
+
+/**
+ * BasePath represents a path sequence in the Firestore database. It is composed
+ * of an ordered sequence of string segments.
+ *
+ * BasePath is immutable. All mutating operations return new independent
+ * instances.
+ *
+ * ## Subclassing Notes
+ *
+ * BasePath is strictly meant as a base class for concrete implementations. It
```

Comment:
I can think of several ways to implement this:
- FieldPath inherits from BasePath, using CRTP so that methods like Concat can return the correct type. This corresponds well to the current implementation in Objective C and Java. One downside is that the interface is split among classes/files. The other downside is that it's possible to mistakenly use BasePath in a polymorphic context (I tried making it less likely by protecting its constructors/destructor and placing it inside an "impl" namespace);
- FieldPath contains BasePath, which is completely hidden from other code. Avoids all problems from the previous bullet, but requires writing a lot of boilerplate to forward all methods;
- implement all methods as template free functions. This will probably feel awkward to use.

@wilhuff Gil, I remember you mentioned you see the concrete classes as templates. What did you have in mind?